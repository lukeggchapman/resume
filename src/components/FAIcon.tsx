/**
 * FAIcon is an implementation of Font Awesome Free for use with React-PDF.
 *
 * <FAIcon icon="github" fab /> for brand icons.
 * <FAIcon icon="envelope" far /> for regular icons.
 * <FAIcon icon="envelope" fas /> for solid icons.
 */

import React from 'react'
import { Text, StyleSheet } from '@react-pdf/renderer'

import './typography/fonts/fontawesome'
import fa from './typography/fonts/fontawesome/faUnicode.json'

export const styles = StyleSheet.create({
  icon: {
    fontFamily: 'Font Awesome 5 Free',
  },
  fab: {
    fontWeight: 700,
  },
  far: {
    fontWeight: 400,
  },
  fas: {
    fontWeight: 900,
  },
})

export type IconStyleProps =
  | {
      fab: true
      far?: false
      fas?: false
    }
  | {
      fab?: false
      far: true
      fas?: false
    }
  | {
      fab?: false
      far?: false
      fas: true
    }

export interface FAIconPropsCommon {
  icon: keyof typeof fa
}
export type FAIconProps = FAIconPropsCommon & IconStyleProps

const getTrueKey = (obj: IconStyleProps) => {
  const keys = (Object.keys(obj) as unknown) as (keyof IconStyleProps)[]

  return keys.find(key => !!obj[key]) as keyof IconStyleProps
}

const FAIcon = ({ icon, ...iconStyleProps }: FAIconProps) => {
  const faStyle = styles[getTrueKey(iconStyleProps)]

  return <Text style={[styles.icon, faStyle]}>{fa[icon]}</Text>
}

export default FAIcon
