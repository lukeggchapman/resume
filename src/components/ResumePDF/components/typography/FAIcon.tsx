/**
 * FAIcon is an implementation of Font Awesome Free for use with React-PDF.
 *
 * <FAIcon icon="github" fab /> for brand icons.
 * <FAIcon icon="envelope" far /> for regular icons.
 * <FAIcon icon="envelope" fas /> for solid icons.
 */

import React from 'react'
import { Text, StyleSheet } from '@react-pdf/renderer'

import './fonts/fontawesome'
import fa from './fonts/fontawesome/faUnicode.json'

export const styles = StyleSheet.create({
  icon: {
    fontFamily: 'Font Awesome 5 Free',
  },
  fab: {
    fontWeight: 700,
  },
  far: {
    fontWeight: 400,
  },
  fas: {
    fontWeight: 900,
  },
})

export type IconStyleProps =
  | {
      fab: true
      far?: false
      fas?: false
    }
  | {
      fab?: false
      far: true
      fas?: false
    }
  | {
      fab?: false
      far?: false
      fas: true
    }

export interface FAIconProps {
  icon: keyof typeof fa
}

export default ({ icon, fab, far, fas }: FAIconProps & IconStyleProps) => {
  const faStyle =
    (fab && styles.fab) ||
    (far && styles.far) ||
    (fas && styles.fas) ||
    styles.far

  return <Text style={[styles.icon, faStyle]}>{fa[icon]}</Text>
}
