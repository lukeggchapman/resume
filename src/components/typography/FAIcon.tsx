/**
 * FAIcon is an implementation of Font Awesome Free for use with React-PDF.
 *
 * <FAIcon icon="github" fab /> for brand icons.
 * <FAIcon icon="envelope" far /> for regular icons.
 * <FAIcon icon="envelope" fas /> for solid icons.
 */

import React from 'react'
import { Text, StyleSheet } from '@react-pdf/renderer'

import './fonts/fontawesome'
import fa from './fonts/fontawesome/faUnicode.json'

export const styles = StyleSheet.create({
  icon: {
    fontFamily: 'Font Awesome 5 Free',
  },
  fab: {
    fontWeight: 700,
  },
  far: {
    fontWeight: 400,
  },
  fas: {
    fontWeight: 900,
  },
})

export type IconStyleProps =
  | {
      fab: true
      far?: false
      fas?: false
    }
  | {
      fab?: false
      far: true
      fas?: false
    }
  | {
      fab?: false
      far?: false
      fas: true
    }

export interface FAIconProps {
  icon: keyof typeof fa
}

const getTrueKey = (obj: IconStyleProps) => {
  const keys = (Object.keys(obj) as unknown) as (keyof IconStyleProps)[]

  return keys.find(key => !!obj[key]) as keyof IconStyleProps
}

export default ({ icon, ...iconStyleProps }: FAIconProps & IconStyleProps) => {
  const faStyle = styles[getTrueKey(iconStyleProps)]

  return <Text style={[styles.icon, faStyle]}>{fa[icon]}</Text>
}
