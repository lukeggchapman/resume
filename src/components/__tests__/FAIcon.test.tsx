import React from 'react'
import { shallow } from 'enzyme'

import FAIcon, { FAIconPropsCommon, styles } from '../FAIcon'

const baseProps: FAIconPropsCommon = {
  icon: 'envelope',
}

describe('FAIcon', () => {
  it('renders correctly', () => {
    expect(shallow(<FAIcon {...baseProps} far />)).toMatchSnapshot()
  })

  it('supports fab styles', () => {
    const faIcon = shallow(<FAIcon {...baseProps} fab />)

    expect(faIcon.prop('style')).toContain(styles.fab)
  })

  it('supports far styles', () => {
    const faIcon = shallow(<FAIcon {...baseProps} far />)

    expect(faIcon.prop('style')).toContain(styles.far)
  })

  it('supports fas styles', () => {
    const faIcon = shallow(<FAIcon {...baseProps} fas />)

    expect(faIcon.prop('style')).toContain(styles.fas)
  })
})
