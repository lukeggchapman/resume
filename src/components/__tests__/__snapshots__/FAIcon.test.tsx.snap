// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAIcon renders correctly 1`] = `
<Text
  style={
    [
      {
        "fontFamily": "Font Awesome 5 Free",
      },
      {
        "fontWeight": 400,
      },
    ]
  }
>
  ïƒ 
</Text>
`;
